ignite/metrics/mutual_information.py

```diff
@@ -0,0 +1,109 @@
+from typing import Sequence
+
+import torch
+import torch.nn.functional as F
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import Metric, reinit__is_reduced, sync_all_reduce
+
+__all__ = ["MutualInformation"]
+
+
+class MutualInformation(Metric):
+    r"""Calculates the `mutual information <https://en.wikipedia.org/wiki/Mutual_information>`_
+    between input :math:`X` and prediction :math:`Y`.
```

Comment:
What is X here ? I'm a bit confused about the definition of MutualInformation here in the docstring, the implementation that just uses predictions and the fact that we can't match sklearn mutual info score: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mutual_info_score.html#sklearn.metrics.mutual_info_score

It would be better to provide metrics that match sklearn metrics instead of something else

You are saying in the description that : 
>  added MutualInformation metric, which is often used in representation learning.

can you please provide any links where it is used and the definition they have ? Thanks!

Reply:
X is an input and model predictions correspond to the posterior distribution $P(Y|X)$, which includes information of X.
Thus, we can compute the mutual information between X and Y just using the predictions.

This mutual information is also sometimes called infomax loss.
For instance, it is used in the followings:
- https://proceedings.neurips.cc/paper/1991/hash/a8abb4bb284b5b27aa7cb790dc20f80b-Abstract.html (NeurIPS1991) Eqs. (7) - (9)
- https://papers.nips.cc/paper_files/paper/2010/hash/42998cf32d552343bc8e460416382dca-Abstract.html (NeurIPS2010) Eq. (1)
- https://arxiv.org/abs/1206.6438 (ICML2012) Eq. (5)
- https://proceedings.mlr.press/v70/hu17b.html (ICML2017) Eqs. (1) and (7)

Reply:
Thanks for the links @kzkadc !
I see now, docstring definitions for H(Y) and H(Y|X) are coming from (8), (9) of this paper: http://proceedings.mlr.press/v70/hu17b/hu17b.pdf "Learning Discrete Representations via Information Maximizing Self-Augmented Training".

Can you please put a ref to it into the docstring, otherwise I can't find a similar defs on the wikipedia page ?